### PR TITLE
Correctly mark promoted extensions as supported

### PR DESCRIPTION
**Pull request by ShabbyX**
`WrappedVulkan::vkCreateInstance` sets instance extension support based
on whether:

- Extension is enabled
- Instance version >= core version in which ext is promoted
- etc

However, the code that looped through the extensions was inside a loop
over the enabled extensions.  If there are no enabled extensions, this
led to extensions promoted to core never being inspected.

The code is split such that enabled extensions are checked within a
loop, and promoted extensions are separately checked.

### CODE REVIEW

**Comment by ShabbyX** [comment]
Noticed when trying to use `vkGetPhysicalDeviceProperties2` without enabling `VK_KHR_get_physical_device_properties2` on the device, even though the Vulkan version is 1.3. RenderDoc never loaded the entry point because the ext is not enabled.

**Comment by baldurk** [comment]
Without a more specific explanation this does not look correct. No functions should be unconditionally returned aside from those from core 1.0, all other functions are conditional on extension or core version enablement (which I handle via implicitly considering promoted extensions enabled on a core version being enabled).

I wrote a simple test and calling `vkGetPhysicalDeviceProperties2` on a vulkan 1.3 instance without the extension works as expected for me:

```
#include <stdio.h>
#include <vulkan/vulkan.h>

int main(int argc, char **argv)
{
  uint32_t ver = VK_MAKE_VERSION(1, 3, 0), devCount = 8;
  VkApplicationInfo app = {VK_STRUCTURE_TYPE_APPLICATION_INFO, NULL, "a", ver, "a", ver, ver};
  VkInstanceCreateInfo instInfo = {VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO, NULL, 0, &app};

  VkInstance inst;
  VkPhysicalDevice physDevices[8] = {};
  vkCreateInstance(&instInfo, NULL, &inst);

  PFN_vkGetPhysicalDeviceProperties2 gpdp2 =
      (PFN_vkGetPhysicalDeviceProperties2)vkGetInstanceProcAddr(inst,
                                                                "vkGetPhysicalDeviceProperties2");

  vkEnumeratePhysicalDevices(inst, &devCount, physDevices);

  VkPhysicalDeviceProperties2 props = {VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PROPERTIES_2};
  gpdp2(physDevices[0], &props);

  while(true)
    fprintf(stderr, "dev 0 is %s\n", props.properties.deviceName);

  return 0;
}
```

Prints the device name on its own and when run through RenderDoc.

If you think this is a bug please provide a reproducible test case and explanation of what exactly is wrong.

**Comment by ShabbyX** [comment]
I'll see about reducing this to a specific program, but in the meantime, what I observed is that the app is creating a Vulkan instance with Version 1.1 (I was wrong, 1.3 was on desktop, and RenderDoc worked fine) without enabling `VK_KHR_get_physical_device_properties2`, and then gets `vkGetPhysicalDeviceProperties2` which is non-null, but calling it crashed inside RenderDoc (in `WrappedVulkan::vkGetPhysicalDeviceProperties2`) because it never initialized the entry point function pointer (`ObjDisp(physicalDevice)->GetPhysicalDeviceProperties2` was nullptr). The simple act of enabling the ext in `VkInstanceCreateInfo` made RenderDoc load the entry point, which led me to this PR.

> all other functions are conditional on extension or core version enablement (which I handle via implicitly considering promoted extensions enabled on a core version being enabled)

Where is that happening? It looks like `CheckExt` is supposed to be doing that, but it's a little hard to follow. I see that happening in `WrappedVulkan::Initialise`, but my logs showed that in `InitDeviceExtensionTables`, `KHR_get_physical_device_properties2` was false.

Thanks for pointing that out though, I'll dig a bit deeper to see why that flag is not set despite Vulkan 1.1.

**Comment by ShabbyX** [comment]
I traced the issue to `WrappedVulkan::vkCreateInstance`:

```
  for(uint32_t i = 0; i < modifiedCreateInfo.enabledExtensionCount; i++)
  {
    CheckInstanceExts();
  }
```

This only checks the extensions if there is at least one extension enabled. In your test, I bet one of `VK_EXT_DEBUG_UTILS_EXTENSION_NAME` or `VK_EXT_DEBUG_REPORT_EXTENSION_NAME` is added to `modifiedCreateInfo`, so `modifiedCreateInfo.enabledExtensionCount` is 1. On my phone, for some reason that's not happening and `modifiedCreateInfo.enabledExtensionCount` is 0.

However, the fact that extension X is supported due to Vulkan 1.1 has nothing to do with whether extension Y is being enabled. I split the `CheckExt` logic here so that enabled extensions and promoted extensions are separately checked.

**Review by baldurk — APPROVED** [review]
Thanks for digging deeper into this, that makes a lot more sense now. I guess something must be happening internally somewhere for the snippet I posted to add an extension or somehow work around this bug, because it did start crashing on me when I lowered the instance version to 1.0.

Regardless this is clearly broken and your fix is appropriate, so it doesn't matter either way.

**Comment by ShabbyX** [comment]
> I guess something must be happening internally somewhere

It must be this, but I haven't looked into it beyond noticing it as the only thing that adds extensions:

```
    // always enable debug report/utils, if it's available
    if(!hasDebugUtils)
    {
      for(const VkExtensionProperties &ext : supportedExts)
      {
        if(!strcmp(ext.extensionName, VK_EXT_DEBUG_UTILS_EXTENSION_NAME))
        {
          addedExts[modifiedCreateInfo.enabledExtensionCount++] = VK_EXT_DEBUG_UTILS_EXTENSION_NAME;
          break;
        }
      }
    }
    else if(!hasDebugReport)
    {
      for(const VkExtensionProperties &ext : supportedExts)
      {
        if(!strcmp(ext.extensionName, VK_EXT_DEBUG_REPORT_EXTENSION_NAME))
        {
          addedExts[modifiedCreateInfo.enabledExtensionCount++] = VK_EXT_DEBUG_REPORT_EXTENSION_NAME;
          break;
        }
      }
    }
```
